File: halopsa-samples/06-inbound-python/halo_to_cf_inbound/app/webhook/change_tracker.py
"""In-memory change tracker, same shape as the CW inbound sample."""

from __future__ import annotations

from dataclasses import dataclass
from threading import Lock


@dataclass(frozen=True)
class TicketState:
    last_action_id: int
    closed: bool


class ChangeTracker:
    def __init__(self) -> None:
        self._state: dict[int, TicketState] = {}
        self._lock = Lock()

    def snapshot(self, ticket_id: int) -> TicketState:
        with self._lock:
            return self._state.get(ticket_id, TicketState(last_action_id=0, closed=False))

    def record(self, ticket_id: int, state: TicketState) -> None:
        with self._lock:
            self._state[ticket_id] = state

File: halopsa-samples/04-inbound-webapp/ContraForce.Samples.HaloInbound/Webhook/ChangeTracker.cs
using System.Collections.Concurrent;

namespace ContraForce.Samples.HaloInbound.Webhook;

/// <summary>
/// Tracks the last processed ticket state per Halo ticket id so that repeated
/// webhook deliveries don't generate duplicate comments on the ContraForce
/// side. Halo fires a webhook on every meaningful change, similar to CW.
/// </summary>
/// <remarks>
/// In-memory only. For multi-instance deployments, replace with Redis,
/// Cosmos, or a SQL-backed implementation.
/// </remarks>
public sealed class ChangeTracker
{
    private readonly ConcurrentDictionary<int, TicketState> _state = new();

    public TicketState Snapshot(int ticketId) =>
        _state.TryGetValue(ticketId, out var s)
            ? s
            : new TicketState(LastActionId: 0, Closed: false);

    public void Record(int ticketId, TicketState state) => _state[ticketId] = state;
}

public sealed record TicketState(int LastActionId, bool Closed);
